### PR TITLE
Milx KML export

### PR DESCRIPTION
**Pull request by domi4484**

### CODE REVIEW

File: kadas/app/milx/kadasmilxexportdialog.cpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include <qgis/qgsproject.h>
+
+#include "kadasapplication.h"
+#include "kadasmainwindow.h"
+#include "kadas/gui/milx/kadasmilxlayer.h"
+#include <milx/kadasmilxexportdialog.h>
```
Comment (m-kuhn) [inline kadas/app/milx/kadasmilxexportdialog.cpp:22]:
```suggestion
#include "milx/kadasmilxexportdialog.h"
```

this is an internal header (not from an external lib)

File: kadas/app/milx/kadasmilxexportdialog.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      ui.listWidget->addItem( item );
+      ui.comboCartouche->addItem( layer->name(), layer->id() );
+    }
+  }
+}
+
+KadasMilxExportDialog::~KadasMilxExportDialog()
+{
+
+}
```
Comment (m-kuhn) [inline kadas/app/milx/kadasmilxexportdialog.cpp:49]:
```suggestion
KadasMilxExportDialog::~KadasMilxExportDialog() = default;
```

is this even required?

File: kadas/app/milx/kadasmilxintegration.cpp [hunk 3]
```diff
@@ -412,6 +390,168 @@ void KadasMilxIntegration::saveMilxly()
...
+        milxLayerEl.appendChild( cartoucheDoc.documentElement() );
+      }
+    }
+  }
+  QString inputXml = doc.toString();
+
+  qDebug() << "inputXml";
+  qDebug() << inputXml;
+
+  qDebug() << "-----------------------------------------------------------------------------------";
```
Comment (m-kuhn) [inline kadas/app/milx/kadasmilxintegration.cpp:515]: I guess we want to get rid of some debugs before merging?

File: kadas/app/milx/kadasmilxintegration.cpp [hunk 4]
```diff
@@ -412,6 +390,168 @@ void KadasMilxIntegration::saveMilxly()
...
+    filename += ".kmz";
+  }
+  else if ( selectedFilter == filters[1] && !filename.endsWith( ".kml", Qt::CaseInsensitive ) )
+  {
+    filename += ".kml";
+  }
+  QString currentVersionTag; KadasMilxClient::getCurrentLibraryVersionTag( currentVersionTag );
+
+  QIODevice *dev = 0;
+  QuaZip *zip = 0;
```
Comment (m-kuhn) [inline kadas/app/milx/kadasmilxintegration.cpp:431]:
```suggestion
  std::unique_ptr<QIODevice> dev;
  std::unique_ptr<QuaZip> zip;
```

could we get rid of manual memory management here?

File: kadas/app/milx/kadasmilxintegration.cpp [hunk 5]
```diff
@@ -412,6 +390,168 @@ void KadasMilxIntegration::saveMilxly()
...
+  if ( selectedFilter == filters[0] )
+  {
+    zip = new QuaZip( filename );
+    zip->open( QuaZip::mdCreate );
+    dev = new QuaZipFile( zip );
+    QuaZipNewInfo info( "Layer.kml" );
+    info.setPermissions( QFile::ReadOwner | QFile::ReadUser | QFile::ReadGroup | QFile::ReadOther );
+    static_cast<QuaZipFile *>( dev )->open( QIODevice::WriteOnly, info );
```
Comment (m-kuhn) [inline kadas/app/milx/kadasmilxintegration.cpp:439]:
```suggestion
    QuaZipFile *zipDev = new QuaZipFile( zip );
    QuaZipNewInfo info( "Layer.kml" );
    info.setPermissions( QFile::ReadOwner | QFile::ReadUser | QFile::ReadGroup | QFile::ReadOther );
    zipDev->open( QIODevice::WriteOnly, info );
    dev.reset( zipDev );
```

File: kadas/app/milx/kadasmilxintegration.cpp [hunk 6]
```diff
@@ -412,6 +390,168 @@ void KadasMilxIntegration::saveMilxly()
...
+  }
+  else
+  {
+    dev = new QFile( filename );
```
Comment (m-kuhn) [inline kadas/app/milx/kadasmilxintegration.cpp:443]:
```suggestion
    dev = std::make_unique<QFile>( filename );
```

**Comment by m-kuhn** [comment]
Thanks @domi4484 I did a quick technical review on this one, didn't test the UI at all.

File: kadas/app/milx/kadasmilxexportdialog.cpp (hunk 1 above)
Comment (domi4484) [reply kadas/app/milx/kadasmilxexportdialog.cpp:22]: thanks too many copy paste, will search the others to

File: kadas/app/milx/kadasmilxexportdialog.cpp (hunk 2 above)
Comment (domi4484) [reply kadas/app/milx/kadasmilxexportdialog.cpp:49]: can be removed completely

File: kadas/app/milx/kadasmilxintegration.cpp (hunk 3 above)
Comment (domi4484) [reply kadas/app/milx/kadasmilxintegration.cpp:515]: changed to draft

File: kadas/app/milx/kadasmilxexportdialog.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+KadasMilxExportDialog::KadasMilxExportDialog( QWidget *parent )
+  : QDialog( parent )
+{
+  ui.setupUi( this );
+
+  setExportMode( ExportMode::Milx );
+
+  ui.comboCartouche->addItem( tr( "Don't add" ) );
+
+  for ( QgsMapLayer *layer : QgsProject::instance()->mapLayers().values() )
```
Comment (m-kuhn) [inline kadas/app/milx/kadasmilxexportdialog.cpp:33]:
```suggestion
  const QList<QgsMapLayer> layers = QgsProject::instance()->mapLayers().values();
  for ( QgsMapLayer *layer : layers )
```

@3nids you also adjusted that in a last PR?

File: kadas/app/milx/kadasmilxintegration.cpp [hunk 8]
```diff
@@ -412,6 +390,127 @@ void KadasMilxIntegration::saveMilxly()
...
+
+  KadasMilxExportDialog exportDialog;
+  exportDialog.setExportMode( KadasMilxExportDialog::ExportMode::Kml );
+
+  if ( exportDialog.exec() == QDialog::Rejected )
+  {
+    return;
+  }
+
+  QStringList exportLayers = exportDialog.selectedLayers();
```
Comment (m-kuhn) [inline kadas/app/milx/kadasmilxintegration.cpp:405]:
```suggestion
  const QStringList exportLayers = exportDialog.selectedLayers();
```

Avoids a detach in the loop IIRC

**Review by m-kuhn — APPROVED** [review]
LGTM code wise
In general it would be nice to make more use of `const` vars, but that's just nitpicking

**Comment by domi4484** [comment]
> LGTM code wise In general it would be nice to make more use of `const` vars, but that's just nitpicking

thanks for the hint, is it something that I never think about

**Pull request merged**